### DISCUSSION TITLE
Some CSS properties are not imported correctly after App re-import

### DISCUSSION THREAD

**Discussion by sergey-dme** [Budibase/budibase#2160] [Help]
Hello @aptkingston and all,

Well.. what I was afraid of (when updating bb every time) is happened again.

I am very excited about new Conditional UI functionality. 
Unfortunately it does not work at all in my existing app. I decided - No problem, let's try to re-import the app! 
And surprises with UI and styles that I am afraid of happened again. 
Link components styling (hover) is replaced now after app re-import to different one, and I'm unable to revert it back/change. My app UI is not the same now, even existing one (backdrop stopped working in screens) and I am still unable to start testing it with my colleagues next week as I was planning :( 
At least my buttons colors on hover were not replaced.

![image](https://user-images.githubusercontent.com/53024832/127377949-ec636df1-48c2-4d7c-aef9-3845dec88619.png)

Can you please do not overwrite/replace exiting components CSS properties that user created in previous versions?
Specially hover. Please 🙏

Conditional UI is working great in re-imported app. 
But during the app re-import I had JSON Error message. I don't know where is log located to see what happened.

Thank you.

**Comment by sergey-dme** [comment]
There is one more thing that made me crying: 
Custom CSS: `backdrop-filter: saturate(180%) blur(20px);` I used it on some containers and it stopped working after BB update, even in existing app.... 

@aptkingston this CSS code it is a key element of design on many my screens (to use opacity and blur fx on containers when bg pictures behind). 
What should I do now? Will the fix released? I love BB, spent two 2.5 months for my app and hoped to start use on next week it with my colleagues but I don't want to change my whole UI I did and I like.

`backdrop-filter: saturate(180%) blur(20px);` works on containers in layouts but not in Screens.

Also some white pixel border appeared after app re-import:

![image](https://user-images.githubusercontent.com/53024832/127382181-90c92f75-ebff-47fe-a7f5-a1c4dd2f559a.png)

I don't know how to fix this white pixel border. It's on every container.
It is more noticeable on containers with bg and if they have 100% page width.

I will be staying on my existing app waiting the updates. Re-design whole my app and removing lots of details would be nightmare for me - I already recently did it when imported in v.0.9.78

I don't want to think I wasted time creating my app, I still hope to start working in the app with colleagues.

**Reply by shogunpurple** [reply]
Hi @sergey-dme I've just pushed a fix for this, that will remove the border you are seeing. You will have to upgrade your client library, which you can do from the client lib update button that looks like this:
<img width="47" alt="Screenshot 2021-07-28 at 22 51 21" src="https://user-images.githubusercontent.com/11256663/127400915-629b4ec0-2fd8-40b1-bf9a-75806ad47e6a.png">

The release will take about 10-15 mins to roll out, but after try to update and let me know if that works for you!

**Reply by sergey-dme** [reply]
@shogunpurple I highly appreciated such quick fix pushing, thank you very much and God bless you!
I have a question regarding client library upgrade - as I understand, it allows to use new features/functionality without re-importing app, but can I revert it back in case some issues with UI like these happened? 

Currently, I keep my existing app without upgrading client library just in case and because UI is not broken there. 

This existing app I re-imported in v.0.9.78 fixed something, specially updated new layouts with new containers types and now it like a baseline-version for me. I afraid to upgrade it and lost elements  in case UI issues... That's why I tested every update in new re-imported app, keeping existing one safe.

Thanks again

**Reply by shogunpurple** [reply]
> I have a question regarding client library upgrade - as I understand, it allows to use new features/functionality without re-
> importing app, but can I revert it back in case some issues with UI like these happened?

That's correct! 

> This existing app I re-imported in v.0.9.78 fixed something, specially updated new layouts with new containers types and > now it like a baseline-version for me. I afraid to upgrade it and lost elements in case UI issues... That's why I tested every > update in new re-imported app, keeping existing one safe.

That's a good plan. We do plan to provide more sophisticated backup/restore mechanisms in future, but this is the best way to manage it now - keep a baseline version, export/import, and make sure everything works, before making the changes on your baseline version.

**Comment by sergey-dme** [comment]
Update: that CSS code works on new containers, but not on existing ones. Still investigating

**Reply by joebudi** [reply]
Hey @sergey-dme

Apologies for the design issues. Is the blur filter not working, saturate, both? Is the image appearing?

**Reply by sergey-dme** [reply]
@joebudi Thank you for your reply. 

@aptkingston @joebudi  So, I tried to re-import my app two more times:

**Good news:**

- No JSON Error message appeared. 
- Custom CSS `backdrop-filter: saturate(180%) blur(20px); `works everywhere (Thanks God) 
_(I think it might be Chrome Issue on my office computer on Windows, I am on mac currently at home and it works fine in both apps in Chrome.)_

**Bad news:**

- Still 1px white unwanted padding (or border) almost everywhere appeared in imported app (screen attached)
- My Links hover ex. CSS properties replaced by default ones (blue)
- Some headlines color broken and positions shifted. 
For example, the text color replaced to black one and some elements now became very ugly (my breadcrumbs as example)
(I could fix shifted position and color but it is for every step on 5 screens!)

![BreadCrumbs](https://user-images.githubusercontent.com/53024832/127395282-b0dfe693-fd3c-4dfc-9c6b-1bf63ba4d616.png)

![Border-Chrome](https://user-images.githubusercontent.com/53024832/127395482-e799e87c-bc5d-4c05-a3f0-36f162d7d1be.png)

The color also changed everywhere in print forms:
(black on grey when it is originally white on grey in existing app)

<img width="983" alt="Screenshot 2021-07-29 at 00 36 05" src="https://user-images.githubusercontent.com/53024832/127399507-3987a634-847b-42ba-8b14-0209941a01a4.png">

**Reply by sergey-dme** [reply]
Update: the `backdrop-filter: saturate(180%) blur(20px);`  CSS stopped working on containers in Screens (only works in Layouts) on Chrome Windows. But on Chrome Mac it works everywhere.

**Reply by sergey-dme** [reply]
Jus an example, Chrome Windows:

![Blur-notworking](https://user-images.githubusercontent.com/53024832/127398313-6b50aec6-b529-43e8-af8a-2419f6a536c4.PNG)

Chrome Mac:

<img width="662" alt="Screenshot 2021-07-29 at 00 24 10" src="https://user-images.githubusercontent.com/53024832/127398196-a630673e-0b15-4124-b200-7121abe0a5de.png">

P.s. bg position is different due to viewport size

**Reply by shogunpurple** [reply]
@sergey-dme the links that you have there, is that just the standard breadcrumb component that comes with our auto screens, but you have styled them to look different?

**Reply by sergey-dme** [reply]
@shogunpurple You made one?? Wow not yet checked. No, I created this from a lot of containers, headlines and custom CSS of course. Will check new one you talking about.. But anyway I hope to use mine.. I like its style )

**Reply by sergey-dme** [reply]
@shogunpurple Oh I misunderstood, you talking about links on my custom Header/Nav - they were created/customized in old builder, correctly imported in 0.9.78, but in 0.9.84 hover color replaced to unwanted blue one.

I thought you talked above about breadcrumbs/wizard, that is not yet released

**Comment by sergey-dme** [comment]
Same app, same element and CSS:

Chrome WIndows (not working):

![win](https://user-images.githubusercontent.com/53024832/127398808-d07e8972-f774-4108-a0fd-6c753c6a4fc4.PNG)

Chrome Mac (working):

<img width="907" alt="Screenshot 2021-07-29 at 00 28 45" src="https://user-images.githubusercontent.com/53024832/127398876-249cf80f-3c47-4ccf-8b39-2c9360e84f92.png">


P.S. On Chrome WIndows it starts working only for new created containers.

**Reply by shogunpurple** [reply]
@sergey-dme if you change the custom CSS on the existing container to something else, then revert it back to:

`backdrop-filter: saturate(180%) blur(20px);` does it work?

**Reply by sergey-dme** [reply]
@shogunpurple No, tried many times. Even removed all css for the problematic container and then added it row by row.
I guess it might be because of old transitions applied in old builder versions. Because that backdrop-filterr fx is not working on Chrome Win on containers which had transitions long time ago, when it was available in design section.

**Reply by sergey-dme** [reply]
@shogunpurple As for headlines position shift - It happened because `line-height `CSS property is changed in new version. 
But it is not a big deal for me to fix all of them and "lock" it by adding rows in custom css in existing app, as well as for the colors - to prevent changing them to black when app re-imported. 
Most important for me - retain old hover attributes on links and buttons. It's like a base of app UI... and backdrop fx as well

**Reply by shogunpurple** [reply]
Thank you for the detailed descriptions @sergey-dme - this makes life much easier for us when it comes to resolving these issues. I'll make sure to have a look at the hover attribute issue and get a fix out for that in the morning.

**Reply by sergey-dme** [reply]
@shogunpurple Thank you very much. This fix currently is most anticipated for me. Plans to start testing the app in my department in August come back alive after your help and attention to this.  I already recently re-designed whole app for compatibility with 0.9.78, fixed many things after re-import and don't want to proceed this again and again, moreover some thinhg like hover it's unable to fix by user.

Well, I was about to finish pre-test version, and after the UI fix I am going re-import app and replace a lot of screens using new amazing Conditional UI and then use re-imported app as new baseline version.

**Reply by sergey-dme** [reply]
@shogunpurple Hi Martin, sorry for being annoying but did you have a chance to take a look at the code regarding prevention user-customized hover properties overwriting? 
That is the only issue that prevents me to update the app to use new functionality. The others I can fix.
Hover Bg and color on buttons have not been changed on app re-import, still the same as I did in design section that's okay, but all links are blue on hover after update.
Thank you 🙏

**Reply by shogunpurple** [reply]
Hi @sergey-dme - I did, the link hover style comes directly from the new app theming feature. This means that the links inherit the styling directly from spectrum. We cannot revert back to the old styles, because the idea is that users will simply theme their app, and be able to provide custom theme colours (like link colours) in the admin panel. Until then, I'm trying to figure out a way that would let you just override these styles. I will keep you updated

**Reply by sergey-dme** [reply]
@shogunpurple Okay, Im on standby and very waiting, I don't want to lost UI that I did. For now, al least my buttons properties preserved

**Reply by kevmodrome** [reply]
@sergey-dme There's an escape hatch of sorts that you can use. You can modify the CSS properties that Spectrum (the underlying design system we're using) uses. This won't let you add css or remove css, but you can at least change colors, radius etc. Here's an example I made:

<img width="1519" alt="customcss" src="https://user-images.githubusercontent.com/534488/127535813-0bc98082-1b3a-493b-b31e-16e3e9926c39.png">

You can find the variables in the Adobe Spectrum CSS GitHub repo: https://github.com/adobe/spectrum-css/tree/main/components

For example, the button skin css can be found here: https://github.com/adobe/spectrum-css/blob/main/components/button/skin.css

**Reply by sergey-dme** [reply]
@kevmodrome wow! it seems it is amazing solution for my case. Thanks for that and for the links!  
Im trying to figure out what parameter should I use **to change text color hover on links.**
`--spectrum-button-primary-m-text-color-hover: rgba(0,0,0,.57); `does not work

**Reply by sergey-dme** [reply]
`--spectrum-alias-text-color-hover: rgba(0,0,0,.57); `
`--spectrum-alias-text-color-mouse-focus: `
don't work too

**Reply by kevmodrome** [reply]
> @kevmodrome wow! it seems it is amazing solution for my case. Thanks for that and for the links!
> Im trying to figure out what parameter should I use **to change text color hover on links.**
> `--spectrum-button-primary-m-text-color-hover: rgba(0,0,0,.57); `does not work

You will have to check with the inspector (in your browser) what classes are applied, take a look at those classes and see which custom properties are used.

**Reply by shogunpurple** [reply]
@sergey-dme I should also mention that this is an interim solution, because we will be working on the custom theme variable feature very soon. It would let you change these variable from the builder UI, and have them reflected across your whole app.

**Reply by kevmodrome** [reply]
> `--spectrum-alias-text-color-hover: rgba(0,0,0,.57); `
> `--spectrum-alias-text-color-mouse-focus: `
> don't work too

On my end the first one works fine, not the second one though - I think that's related to another issue though (accessibility)

**Reply by sergey-dme** [reply]
@shogunpurple @kevmodrome okay. I tried many classes on Link component. Goal is to change text color on hover. Nothing work for now...

**Reply by sergey-dme** [reply]
@shogunpurple @kevmodrome  
`--spectrum-global-color-blue-600: rgba(0,0,0,.57); `WORKS!!!

**Reply by sergey-dme** [reply]
Hi @shogunpurple @kevmodrome ` backdrop-filter: saturate(180%) blur(20px);` does not work on any container at all in Screens (even newly created) on Chrome Windows, that only works on containers in Layouts. Would it be possible to fix, please?

**Reply by shogunpurple** [reply]
@sergey-dme we will look into it - seems like it might be a style applied on screen containers directly that isn’t applied in layouts. Thanks!

**Reply by aptkingston** [reply]
Hi @sergey-dme. I've just tested this and `backdrop-filter` works on Chrome on Windows for me both inside screens and inside containers, Here's the quick test I just made:
![image](https://user-images.githubusercontent.com/9075550/127663969-ac26e2ba-eb01-41ac-81ee-cc92373fbc1c.png)

The only custom styles applied to each of those containers is the exact CSS rule you provided. Is your background image at the layout level or inside a screen? Perhaps it depends on where the background is?

**Reply by sergey-dme** [reply]
Hi @aptkingston The full Custom CSS on container:

`min-width: 900px;`
`align-self: center;`
`background: white;`

`box-shadow: rgb(34 42 55 / 4%) 0px 0px 4px, rgb(34 42 55 / 4%) 0px 4px 12px;`
    `border-color: rgba(0, 0, 0, 0.1);`
    `border-width: 2px;`
    `border-style: hidden;`
    `overflow: hidden;`
   ` border-radius: 12px;`

`top: 67px;`
`position: absolute;`

`opacity: 0.9;`
`backdrop-filter: saturate(180%) blur(20px);`

This container on top of another one with img as background (as on my screens above - https://github.com/Budibase/budibase/discussions/2160#discussioncomment-1084465 )
Opacity works, but not backdrop-filter.

It was working on Chrome windows in 0.9.78 before updating to 0.9.84 (now only works on containers in layouts)
_On Chrome Mac it works as before._

**Reply by aptkingston** [reply]
@sergey-dme This is not a problem with Budibase - your CSS is invalid I'm afraid. You have a solid background colour which stops `backdrop-filter` from doing anything. Chrome will not use a filter if you have a solid background.

Please remove `opacity: 0.9` and change `background` to `rgba(255, 255, 255, 0.9)` instead and it will work.

**Reply by sergey-dme** [reply]
@aptkingston Thank you very much and thanks to Martin and Kevin for taking time and help.
Replacement opacity to bg rgba(255, 255, 255, 0.9) solve the problem. But it's interesting that it worked on Chrome Mac, same browser but different implementation depending on OS... did not expect that
